src/k2/dto/SKVRecord.h

```diff
@@ -0,0 +1,172 @@
+/*
+MIT License
+
+Copyright(c) 2020 Futurewei Cloud
+
+    Permission is hereby granted,
+    free of charge, to any person obtaining a copy of this software and associated documentation files(the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and / or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions :
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#pragma once
+
+#include <optional>
+
+#include <k2/dto/FieldTypes.h>
+#include <k2/dto/ControlPlaneOracle.h>
+
+namespace k2 {
+namespace dto {
+
+class SKVRecord {
+public:
+    // The record must be serialized in order. Schema will be enforced
+    template <typename T>
+    void serializeNext(T field) {
+        FieldType ft = TToFieldType<T>();
+        if (fieldCursor >= schema.fields.size() || ft != schema.fields[fieldCursor].type) {
+            throw new std::runtime_error("Schema not followed in record serialization");
+        }
+
+        for (size_t i = 0; i < schema.partitionKeyFields.size(); ++i) {
+            if (schema.partitionKeyFields[i] == fieldCursor) {
+                partitionKeys[i] = FieldToKeyString<T>(field);
+            }
+
+            if (schema.rangeKeyFields[i] == fieldCursor) {
+                rangeKeys[i] = FieldToKeyString<T>(field);
+            }
+        }
+
+        fieldData.write(field);
+        ++fieldCursor;
+    }
+
+    // Skip serializing the next field, for optional fields or partial updates
+    void skipNext();
+
+    // Deserialization can be in any order, but the preferred method is in-order
+    template <typename T>
+    T deserializeField(const String& name) {
+        for (size_t i = 0; i < schema.fields.size(); ++i) {
+            if (schema.fields[i].name == name) {
+                return deserializeField<T>(i);
+            }
+        }
+
+        throw new std::runtime_error("Schema not followed in record deserialization");
+    }
+
+    void seekField(uint32_t fieldIndex);
+
+    template <typename T>
+    T deserializeField(uint32_t fieldIndex) {
```

Comment:
How does this deal with optional fields? I'd expect same response as the deserializeNextOptional() method